solver/llbsolver/vertex.go

```diff
@@ -186,36 +187,102 @@ func newVertex(dgst digest.Digest, op *pb.Op, opMeta *pb.OpMetadata, load func(d
 	return vtx, nil
 }
 
+func recomputeDigests(ctx context.Context, all map[digest.Digest]*pb.Op, visited map[digest.Digest]digest.Digest) error {
+	changed := map[digest.Digest]*pb.Op{}
+	for dgst, op := range all {
```

Comment:
This still seems to cycle over `all` multiple times.

What I thought with the recursive function was something that recomputes the digest for a single `Op` and always returns(and saves to visited) the correct digests with all of its inputs. For `Op` it will first call `recompute` to all of its inputs.

Then there is only one `for` cycle that calls `recompute` for each op. Most cases it will just return the result directly from the `visited` map because op was already processed as a input for another op.